qualtran/surface_code/azure_cost_model.ipynb

```diff
@@ -0,0 +1,708 @@
+   "source": [
+    "from qualtran.surface_code import azure_cost_model\n",
+    "\n",
+    "# First calculate the number of logical qubits (Eq. D1)\n",
+    "logical_qubits = azure_cost_model.logical_qubits(quantum_dynamics_specs)\n",
+    "print('Q =', logical_qubits)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Calculate the minimum number of logical time steps (Eq D3)\n",
+    "error_budget = 0.001\n",
+    "c_min = azure_cost_model.minimum_time_steps(\n",
+    "    error_budget=error_budget, alg=quantum_dynamics_specs, rotation_model=BeverlandEtAlRotationCost\n",
+    ")\n",
+    "print('C_min = %e' % c_min)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# And the number of needed T operations (Eq D4)\n",
+    "t_operations = azure_cost_model.t_states(\n",
+    "    error_budget=error_budget, alg=quantum_dynamics_specs, rotation_model=BeverlandEtAlRotationCost\n",
+    ")\n",
+    "print('M = %e' % t_operations)"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Comparing our esimates of $Q = 230, C_{min} = 1.8 \\times 10^5, M = 6 \\times 10^5$ to the paper estimates of \n",
+    "$Q = 230, C_{min} = 1.5 \\times 10^5, M = 2.4 \\times 10^6$. We find a match for $Q$ and $C_{min}$ however we are off by 4x for the number of T gates $M$.\n",
+    "\n",
+    "D4 gives the formula for $M$ as $$M = M_T + 4 M_{Tof} + M_R \\lceil A \\log_2{M_R/\\epsilon_{syn}} + B\\rceil$$\n",
+    "\n",
+    "Since $M_T = M_{Tof} = 0$, the only contribution to $M$ comes from rotations where $M_R = 30100, \\epsilon_{syn} = \\frac{10^{-3}}{3} , A = 0.53, B = 5.3$ which give our value of 6.02e5."
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Next, we estimate the code distance $d$. $d$ satisfies \n",
+    "$$\n",
+    "P(d) = a \\left ( \\frac{p}{p^*} \\right )^\\frac{d+1}{2} \n",
+    "$$\n",
+    "subject ot the constraint on the logical error rate of $QCP(d) = \\frac{\\epsilon}{3}$. Where $p$ is the physical error rate. $a$ and $p^*$ are constants determined by the QEC scheme."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "d = azure_cost_model.code_distance(\n",
+    "    error_budget=error_budget,\n",
+    "    time_steps=c_min,\n",
+    "    alg=quantum_dynamics_specs,\n",
+    "    qec=BeverlandSuperconductingQubits,\n",
+    "    physical_error_rate=BeverlandEtAl.physical_error_rate,\n",
+    ")\n",
+    "print(f'{d=}')"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Matching the surface code of $d = 9$. This leads to a total run time (Eq. E3) of 0.65s which is close to the time in the paper of 0.55s"
```

Comment:
matching the surface code? Do you mean matching the paper?

Reply:
fixed